### ISSUE TITLE
Complicated incorrect functor application crashes compiler.

### ISSUE THREAD

**Issue by joelatschool** [ocaml/ocaml#12821]
The following code:
```ocaml
module HashPair (T : sig type t val equal : t -> t -> bool end) : Hashtbl.HashedType = struct
  type t = T.t * T.t
  let equal (a, b) (c, d) =
    (T.equal a b) && (T.equal c d) 
  let hash = Hashtbl.hash
end
    
module type Type = sig module type T end
module MakeApply
    (A:Type) (X: A.T)
    (B:Type) (F: A.T -> B.T)
    (C:Type) (H: sig module Make : B.T -> C.T end) = H.Make(F(X))
    
module HashtblPair = MakeApply
    (struct module T = Int end) (*this should be struct module type T = module type of Int *)
    (Int)
    (struct module type T = Hashtbl.HashedType end)
    (HashPair)
    (struct module type T = Hashtbl.S end)
    (Hashtbl)
```
causes a crash on compiler 4.14.1 (but not 5.1.0) with `Fatal error: exception Ctype.Nondep_cannot_erase(_)`.

**Comment by yallop** [comment]
This appears to be the same issue as https://github.com/ocaml/ocaml/issues/12014#issuecomment-1774540161

**Comment by gasche** [comment]
Can the issue be closed, if it is fixed in recent compiler versions, or @joelatschool do you have strong reasons to keep 4.14 and to wish for a fix there as well?

**Comment by joelatschool** [comment]
I don't have any strong reason to keep 4.14, I use OCaml for fun and occasionally for school and anyway I encountered this while trying to recreate an example in a blog post so it's no skin off my nose. I do think there is a case that an issue like this should probably be fixed in 4.14 though given that ocaml.org currently lists it as the current Long Term Support release.

**Comment by Octachron** [comment]
The fix was cherry-picked to the 4.14 branch, and it will be part of the upcoming 4.14.2 release.
